c7n/resources/vpc.py

```diff
@@ -2708,3 +2708,80 @@ class resource_type(query.TypeInfo):
         arn_type = 'traffic-mirror-target'
         universal_taggable = object()
         id_prefix = 'tmt-'
+
+
+@RouteTable.filter_registry.register('cross-az-nat-gateway-route')
+class CrossAZRouteTable(Filter):
+    """Filter route-table which is attached to a subnet that is not in the same availability zone (AZ) of the
+    NAT Gateway subnet AZ. This filter is applicable to those route-tables which have next hop as NatGateways for
+    any route. If a route-table has NAT gateway from us-west-2a and it is associated to subnet from us-west-2b then it
+    will flagged, and NatGatewayInCrossAvailabilityZone key would be set a True.
+
+    :Example:
+    .. code-block:: yaml
+            policies:
+              - name: cross-az-nat-gateway--traffic
+                resource: aws.route-table
+                filters:
+                    - type: cross-az-nat-gateway-route
+                    - type: value
+                      key: Routes[].NatGatewayId
+                      op: eq
+                      value: not-null
+                    - type: value
+                      key: Associations[].NatGatewayAvailableInSubnetAvailabilityZone
+                      op: ni
+                      value: false
+                      value_type: swap
+                    - type: value
+                      key: Associations[].SubnetAvailabilityZone
+                      op: eq
+                      value: not-null
+                    - type: value
+                      key: NatGatewayInCrossAvailabilityZone
+                      op: eq
+                      value: true
+                actions:
+                  - notify
+    """
+    schema = type_schema('cross-az-nat-gateway-route')
+
+    def process(self, resources, event=None):
+        client = local_session(self.manager.session_factory).client('ec2')
+        # dump of all subnets to avoid multiple API calls
+        all_subnets = client.describe_subnets()
+        all_nat_gws = client.describe_nat_gateways()
```

Comment:
@vvishwakarma-intuit results from describe calls are paginated. You'll need to handle pagination.

Reply:
Added pagination as suggested.

Reply:
Added pagination as suggested.